pkgs/top-level/config.nix

```diff
@@ -16,6 +16,18 @@ let
     '');
   });
 
+  overrideFnType = mkOptionType {
+    name = "overrides";
+    check = builtins.isFunction;
+    merge = const
+      (foldl'
+        (final: override: pkgs:
+          let
+            final' = final pkgs;
+          in final' // (override.value (pkgs // final')))
+        (const {}));
+  };
+
```

Comment:
Could you move this to `lib/types.nix` and add a comment like

```nix
# EXPERIMENTAL
# Known issue:
# Overrides composition does not compose symmetrically, exposing the arbitrary
# nature of module `imports` ordering in a potentially harmful manner.
# TODO
#  - Restrict override composition by `throw`-ing when any two layers contain the
#    same attribute name, which may create an `imports` ordering dependency.
#    (a layer is the result of applying one or more override functions)
#  - Apply this restriction only to unordered layers, which means layers that
#    have the same `mkOrder` number.
#  - Document this type in the NixOS manual
```

Moving it allows us to test the code and assert some of its properties.
We already have a few test cases for types in `lib/tests/modules.sh`.
One question that a test case can easily answer is in which order the layers are composed.
It must not be the reverse direction of `overlays`.

Reply:
Those are for overlays (2-arg) where as these only take the `self` argument, no `super`. Seems like that can be fixed with a `map const` to ignore `super`.
I thought a simple reimplementation would be easy to understand, but piggybacking on overlays seems ok.

Reply:
Oh, I mistakenly read `overlayFnType` instead of `overrideFnType` here. I guess this might be fine as is then

Reply:
@roberth @infinisil  I'm happy to implement a better version of that type in the future, but no guarantees that I'll get to it soon :)